File: Day #186/solution.cpp
// Day 186: Minimum subset-sum difference (partition problem).
// Subset-sum DP over total, pick best <= total/2, reconstruct. Time O(n*S), Space O(n*S).
#include <bits/stdc++.h>
using namespace std;

void solve(const vector<int>& a) {
    int n = a.size(), tot = accumulate(a.begin(), a.end(), 0);
    vector<vector<char>> dp(n + 1, vector<char>(tot + 1, 0));
    dp[0][0] = 1;
    for (int i = 1; i <= n; i++)
        for (int s = 0; s <= tot; s++)
            dp[i][s] = dp[i - 1][s] || (s >= a[i - 1] && dp[i - 1][s - a[i - 1]]);
    int best = 0;
    for (int s = tot / 2; s >= 0; s--) if (dp[n][s]) { best = s; break; }
    vector<int> sub, other;
    int s = best;
    for (int i = n; i >= 1; i--) {
        if (s >= a[i - 1] && dp[i - 1][s - a[i - 1]]) { sub.push_back(a[i - 1]); s -= a[i - 1]; }
        else other.push_back(a[i - 1]);
    }
    reverse(sub.begin(), sub.end()); reverse(other.begin(), other.end());
    auto pr = [](const vector<int>& v) {
        cout << "{";
        for (size_t i = 0; i < v.size(); i++) cout << v[i] << (i + 1 < v.size() ? ", " : "");
        cout << "}";
    };
    pr(sub); cout << " and "; pr(other); cout << "\n";
}

int main() {
    solve({5, 10, 15, 20, 25});
    return 0;
}

File: Day #186/solution.rs
// Day 186: Minimum subset-sum difference (partition problem).
// Subset-sum DP over total, pick best <= total/2, reconstruct. Time O(n*S), Space O(n*S).
fn solve(a: &[i64]) {
    let n = a.len();
    let tot: i64 = a.iter().sum();
    let totu = tot as usize;
    let mut dp = vec![vec![false; totu + 1]; n + 1];
    dp[0][0] = true;
    for i in 1..=n {
        for s in 0..=totu {
            dp[i][s] = dp[i - 1][s] || (s >= a[i - 1] as usize && dp[i - 1][s - a[i - 1] as usize]);
        }
    }
    let mut best = 0usize;
    for s in (0..=totu / 2).rev() {
        if dp[n][s] {
            best = s;
            break;
        }
    }
    let mut sub = Vec::new();
    let mut other = Vec::new();
    let mut s = best;
    for i in (1..=n).rev() {
        let ai = a[i - 1] as usize;
        if s >= ai && dp[i - 1][s - ai] {
            sub.push(a[i - 1]);
            s -= ai;
        } else {
            other.push(a[i - 1]);
        }
    }
    sub.reverse();
    other.reverse();
    let f = |v: &[i64]| {
        let parts: Vec<String> = v.iter().map(|x| x.to_string()).collect();
        format!("{{{}}}", parts.join(", "))
    };
    println!("{} and {}", f(&sub), f(&other));
}

fn main() {
    solve(&[5, 10, 15, 20, 25]);
}
